### PR TITLE
Remove Redundant CS Update on Snapshot Finalization

### PR DESCRIPTION
**Pull request by original-brownbear**
This change folds the removal of the in-progress snapshot entry
into setting the safe repository generation. 
Outside of removing an unnecessary cluster state update, this also has the advantage
of removing a somewhat inconsistent cluster state where the safe repository generation points at `RepositoryData` that contains a finished snapshot while it is still in-progress in the cluster
state, making it easier to reason about the state machine of upcoming concurrent snapshot operations.

Note: We can do the same for snapshot deletion where it is a much more interesting change because most of the work in the delete (deleting all unreferenced blobs) can happen after the new index-N is written and does not require the delete-in-progress entry in the cluster state (assuming the repo is using shard-generations). I just wanted to start with the snapshot finalization since it's much easier to review and has no BwC implications.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/discovery/SnapshotDisruptionIT.java [hunk 1]
```diff
@@ -189,8 +189,7 @@ public void clusterChanged(ClusterChangedEvent event) {
                         final RepositoriesMetadata repoMeta =
                             event.state().metadata().custom(RepositoriesMetadata.TYPE);
                         final RepositoryMetadata metadata = repoMeta.repository("test-repo");
-                        if (metadata.generation() == metadata.pendingGeneration()
-                            && metadata.generation() > snapshotEntry.repositoryStateId()) {
+                        if (metadata.pendingGeneration() > snapshotEntry.repositoryStateId()) {
```
Comment (original-brownbear) [inline server/src/test/java/org/elasticsearch/discovery/SnapshotDisruptionIT.java:192]: The concrete situation that this test was covering is technically gone with this change, but I think it's reasonable to keep testing the very similar situation of master fail-over during the repository side of the finalization here now.

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/2  (some ML thing)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 2]
```diff
@@ -1301,10 +1304,11 @@ public boolean isReadOnly() {
      * @param repositoryData RepositoryData to write
      * @param expectedGen    expected repository generation at the start of the operation
      * @param writeShardGens whether to write {@link ShardGenerations} to the new {@link RepositoryData} blob
+     * @param stateFilter    filter for the last cluster state update executed by this method
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:1307]: It's not a filter, rather a transformation function. Perhaps adapt naming?

**Review by ywelsch — COMMENTED** [review]
I've left one comment about naming. I think that technically this change is ok, but I wonder about the separation of SnapshotsService and BlobStoreRepository (which is only one implementation of Repository), where it becomes less and less clear where the responibilities for each class lie.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 2 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:1307]: Renamed :)

**Comment by original-brownbear** [comment]
> but I wonder about the separation of SnapshotsService and BlobStoreRepository (which is only one implementation of Repository), where it becomes less and less clear where the responibilities for each class lie.

+1 in theory. In practice, we're already exposing a lot of the specific internals of the blob store repository (e.g. shard generations are handled in snapshots-in-progress, repo generation, ...) so that ship seems to have sailed already. Also, practically speaking all our repos are eventually backed by a blob store repository for writing. Maybe it would make more sense to eventually adjust the repository interface to be more centered around just writing blobs and move all the state handling out of it completely. But for now I don't see a short way of enabling more concurrent operations without this kind of mixing of the code.

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/1 (unrelated reindex failure)

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks Yannick!

**Pull request merged**